### PR TITLE
less copying of repeatedly used bind parameter values

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Less copying of repeatedly used bind parameter values in AQL queries.
This reduces memory overhead for queries that access large bind parameter values (large arrays or objects) via repeatedly refering to them by name (e.g. `FOR doc IN collection FILTER doc.value == CONCAT(@largeBind['foo'], @largeBind['bar'], @largeBind[...], ...`).
The bind parameter values are only once converted into an AstNode per query, and from the on only shallow-cloned when accessed.
In some example query this decreased memory usage by 99%.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/BindParameters.cpp [hunk 1]
```diff
@@ -65,28 +66,42 @@ uint64_t BindParameters::hash() const {
   return _builder->slice().hash();
 }
 
-/// @brief mark a bind parameter as "used", and return its value.
+/// @brief return a bind parameter value by name
 /// will return VPackSlice::noneSlice() if the bind parameter does not exist!
-VPackSlice BindParameters::markUsed(std::string const& name) noexcept {
+std::pair<VPackSlice, AstNode*> BindParameters::get(
```
Comment (mpoeter) [inline arangod/Aql/BindParameters.cpp:71]: It no longer returns only a value, so the comment should be updated accordingly.

File: arangod/Aql/BindParameters.cpp [hunk 2]
```diff
@@ -65,28 +66,42 @@ uint64_t BindParameters::hash() const {
...
   if (it == _parameters.end()) {
-    return VPackSlice::noneSlice();
+    return std::make_pair(VPackSlice::noneSlice(), nullptr);
   }
 
-  // mark the bind parameter as being used
-  (*it).second.second = true;
+  // return parameter value and AstNode
+  TRI_ASSERT(!(*it).second.first.isNone());
+  return std::make_pair((*it).second.first, (*it).second.second);
```
Comment (mpoeter) [inline arangod/Aql/BindParameters.cpp:82]: Isn't this just the same as
```suggestion
  return it->second;
```

File: arangod/Aql/Ast.cpp [hunk 3]
```diff
@@ -1888,6 +1911,10 @@ void Ast::injectBindParameters(
             }
           }
           node = newNode;
+
+          if (cachedNode == nullptr) {
+            parameters.registerNode(param, node);
+          }
```
Comment (mpoeter) [inline arangod/Aql/Ast.cpp:1917]: What if `cacheNode != nullptr`? Couldn't we also just lookup the param and use a previously registered datasource node instead?

**Review by mpoeter — APPROVED** [review]
A few small comments, but overall LGTM.

File: arangod/Aql/BindParameters.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/BindParameters.cpp]: Done

File: arangod/Aql/BindParameters.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/BindParameters.cpp]: Absolutely... :facepalm:

File: arangod/Aql/Ast.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/Aql/Ast.cpp]: That _may_ be true, but I am unaware of the side effects for this type of node.
At least it does no harm to always create a new AstNode in this case, as the AstNode will always have a small size (in contrast to arbitrarily-sized bind parameter values from JSON/VPACK).
I added more comments to the code to make the intentions more clear.

Comment (mpoeter) [reply arangod/Aql/Ast.cpp]: Ok... but the newly created nodes should be equivalent to any cached ones. Should we at least add some assertions to check that?

Comment (jsteemann) [reply arangod/Aql/Ast.cpp]: Added assertions in latest commit.

**Pull request merged**